core/src/main/java/io/jenkins/pluginhealth/scoring/probes/IncrementalBuildDetectionProbe.java

```diff
@@ -0,0 +1,147 @@
+        }
+
+        Supplier<Stream<Path>> incrementalConfigs = () -> {
+            try {
+                return Files.find(mvnConfig, 1, (path, $) -> Files.isRegularFile(path) && (path.endsWith("maven.config") || path.endsWith("extensions.xml")));
+            } catch (IOException e) {
+                LOGGER.error("Could not read files from .mvn directory for {} plugin while running {} probe.", plugin.getName(), key());
+                return null;
+            }
+        };
+        Optional<Path> mavenExtensionsFile = incrementalConfigs.get().filter(path -> path.endsWith("extensions.xml")).findFirst();
+        Optional<Path> mavenConfigFile = incrementalConfigs.get().filter(path -> path.endsWith("maven.config")).findFirst();
+
+        if (mavenExtensionsFile.isPresent() && mavenConfigFile.isPresent()) {
+            return isExtensionsXMLConfigured(mavenExtensionsFile.get()) && isMavenConfigConfigured(mavenConfigFile.get())
+                ? ProbeResult.success(key(), String.format("Incremental Build is configured in the %s plugin.", plugin.getName()))
+                : ProbeResult.failure(key(), String.format("Incremental Build is not configured in the %s plugin.", plugin.getName()));
+        }
+        return ProbeResult.failure(key(), String.format("Incremental Build is not configured in the %s plugin.", plugin.getName()));
+    }
+
+    @Override
+    public String[] getProbeResultRequirement() {
+        return new String[]{ContinuousDeliveryProbe.KEY};
+    }
+
+    @Override
+    public String key() {
+        return KEY;
+    }
+
+    @Override
+    public String getDescription() {
+        return "The probe detects whether incremental build is configured in Maven Config.";
+    }
+
+    /**
+     * Checks whether `extensions.xml` is configured in the plugin
+     *
+     * @param path Looks for extensions.xml configuration in the particular path
+     * @return true if a correct configuration is found, otherwise false
+     */
+    private boolean isExtensionsXMLConfigured(Path path) {
+        try {
+            DocumentBuilderFactory dbFactory = DocumentBuilderFactory.newInstance();
+            DocumentBuilder dBuilder = dbFactory.newDocumentBuilder();
+            Document doc = dBuilder.parse(path.toFile());
+            doc.getDocumentElement().normalize();
+            Element rootElement = doc.getDocumentElement();
+            Element extensionElement = (Element) rootElement.getElementsByTagName("extension").item(0);
+            Element groupIdElement = (Element) extensionElement.getElementsByTagName("groupId").item(0);
+            Element artifactIdElement = (Element) extensionElement.getElementsByTagName("artifactId").item(0);
+            String groupId = groupIdElement.getTextContent();
+            String artifactId = artifactIdElement.getTextContent();
+            return groupId.equals(INCREMENTAL_TOOL) && artifactId.equals(INCREMENTAL_TOOL_ARTIFACT_ID);
```

Comment:
Remember that trick Adrien shared with you some time ago?

If `groupId` or `artifactId` are `null`, then calling the `equals` method on them will result in a `NullPointerException`, which you can avoid by simply writing your tests in the other way around, starting with what you know for sure is not `null`:

`INCREMENTAL_TOOL.equals(groupId) && INCREMENTAL_TOOL_ARTIFACT_ID.equals(artifactId)`

Reply:
I will remember more.